lib/scholar/metrics.ex

```diff
@@ -103,42 +162,79 @@ defmodule Scholar.Metrics do
 
   ## Examples
 
-      iex> Scholar.Metrics.recall(Nx.tensor([0, 1, 1, 1]), Nx.tensor([1, 0, 1, 1]))
+      iex> Scholar.Metrics.binary_recall(Nx.tensor([0, 1, 1, 1]), Nx.tensor([1, 0, 1, 1]))
       #Nx.Tensor<
         f32
         0.6666666865348816
       >
 
   """
-  defn recall(y_true, y_pred, opts \\ []) do
+  defn binary_recall(y_true, y_pred, opts \\ []) do
     assert_shape(y_true, Nx.shape(y_pred))
 
-    true_positives = true_positives(y_true, y_pred, opts)
-    false_negatives = false_negatives(y_true, y_pred, opts)
+    true_positives = binary_true_positives(y_true, y_pred, opts)
+    false_negatives = binary_false_negatives(y_true, y_pred, opts)
 
     Nx.divide(true_positives, false_negatives + true_positives + 1.0e-16)
   end
 
+  @doc ~S"""
+  Computes the recall of the given predictions with respect to
+  the given targets for multi-class classification problems.
+
+  ## Argument Shapes
+
+    * `y_true` - $\(d_0, d_1, ..., d_n\)$
+    * `y_pred` - $\(d_0, d_1, ..., d_n\)$
+
+  ## Options
+
+    * `:threshold` - threshold for truth value of the predictions.
+      Defaults to `0.5`
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 1, 1, 1, 1, 0, 2, 1, 0, 1], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 2, 1, 1, 2, 2, 2, 0, 0, 1], type: {:u, 32})
+      iex> Scholar.Metrics.recall(y_true, y_pred, num_classes: 3)
+      #Nx.Tensor<
+        f32[3]
+        [0.6666666865348816, 0.5, 1.0]
+      >
+
+  """
+  defn recall(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, [:num_classes])
+    assert_shape_pattern(y_true, {_})
+    assert_shape(y_pred, Nx.shape(y_true))
+
+    cm = confusion_matrix(y_true, y_pred, opts)
+    true_positive = Nx.take_diagonal(cm)
+    false_negative = Nx.subtract(Nx.sum(cm, axes: [1]), true_positive)
+    Nx.divide(true_positive, true_positive + false_negative + 1.0e-16)
+  end
+
   @doc """
   Computes the number of true positive predictions with respect
-  to given targets.
+  to given targets for multi-class classification problems.
 
   ## Options
 
     * `:threshold` - threshold for truth value of predictions.
       Defaults to `0.5`.
 
+
   ## Examples
 
       iex> y_true = Nx.tensor([1, 0, 1, 1, 0, 1, 0])
       iex> y_pred = Nx.tensor([0.8, 0.6, 0.4, 0.2, 0.8, 0.2, 0.2])
-      iex> Scholar.Metrics.true_positives(y_true, y_pred)
+      iex> Scholar.Metrics.binary_true_positives(y_true, y_pred)
       #Nx.Tensor<
         u64
         1
       >
   """
-  defn true_positives(y_true, y_pred, opts \\ []) do
+  defn binary_true_positives(y_true, y_pred, opts \\ []) do
```

Comment:
We don't need these either right? We can keep them as private if they are useful.

We may want to benchmark functions like binary precision/recall using this implementation vs using confusion matrix, but we can do that separately.

Also, I wonder if accepting `:threshold` everywhere makes sense, given that we always just compute `thresholded_preds`. This could very well be a separate function, like one-hot conversion etc.

Reply:
I think for binary cases, current implementation should be faster because `confusion_matrix` will first construct an `{n, 2}`-shape indices tensor and then update the confusion matrix `n` times.

Reply:
> then update the confusion matrix n times

Pretty sure `indexed_add` is more optimized than that! Each individual tensor operation is closer to "all at once", and XLA optimizes across operations via fusion and whatnot, so it's best to measure various options.

Reply:
Oh I didn't take parallelised operation into consideration! Then `indexed_add` should be fairly fast. Let's see the benchmark data and decide. :)